File: src/main/python/AoC2024_04.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 4
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.geometry import Direction
from aoc.grid import CharGrid

Input = CharGrid
Output1 = int
Output2 = int


TEST = """\
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return CharGrid.from_strings(list(input_data))

    def part_1(self, grid: Input) -> Output1:
        return sum(
            all(
                (n := next(it, None)) is not None and grid.get_value(n) == v
                for v in ["M", "A", "S"]
            )
            for it in (
                grid.get_cells_dir(cell, d)
                for cell in grid.get_all_equal_to("X")
                for d in Direction.octants()
            )
        )

    def part_2(self, grid: Input) -> Output2:
        dirs = [
            Direction.LEFT_AND_UP,
            Direction.RIGHT_AND_DOWN,
            Direction.RIGHT_AND_UP,
            Direction.LEFT_AND_DOWN,
        ]
        matches = {"MSMS", "SMSM", "MSSM", "SMMS"}
        return sum(
            grid.get_value(cell) == "A"
            and "".join(grid.get_value(cell.at(d)) for d in dirs) in matches
            for cell in grid.get_cells_without_border()
        )

    @aoc_samples(
        (
            ("part_1", TEST, 18),
            ("part_2", TEST, 9),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_04.java
import static com.github.pareronia.aoc.Utils.toAString;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_04
        extends SolutionBase<CharGrid, Integer, Integer> {
    
    private AoC2024_04(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_04 create() {
        return new AoC2024_04(false);
    }
    
    public static AoC2024_04 createDebug() {
        return new AoC2024_04(true);
    }
    
    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(inputs);
    }
    
    @Override
    public Integer solvePart1(final CharGrid grid) {
        return (int) grid.getAllEqualTo('X')
            .flatMap(cell -> Direction.OCTANTS.stream()
                                .map(d -> grid.getCells(cell, d).iterator()))
            .filter(it -> Stream.of('M', 'A', 'S').allMatch(
                    ch -> it.hasNext() && grid.getValue(it.next()) == ch))
            .count();
    }
    
    @Override
    public Integer solvePart2(final CharGrid grid) {
        final List<Direction> dirs = List.of(
            Direction.LEFT_AND_UP,
            Direction.RIGHT_AND_DOWN,
            Direction.RIGHT_AND_UP,
            Direction.LEFT_AND_DOWN
        );
        final Set<String> matches = Set.of("MSMS", "SMSM", "MSSM", "SMMS");
        return (int) grid.getCellsWithoutBorder()
            .filter(cell -> grid.getValue(cell) == 'A')
            .filter(cell -> matches.contains(
                    dirs.stream()
                        .map(d -> grid.getValue(cell.at(d)))
                        .collect(toAString())))
            .count();
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "18"),
        @Sample(method = "part2", input = TEST, expected = "9"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_04.create().run();
    }

    private static final String TEST = """
            MMMSXXMASM
            MSAMXMSMSA
            AMXSXMAAMM
            MSAMASMSMX
            XMASAMXAMM
            XXAMMXXAMA
            SMSMSASXSS
            SAXAMASAAA
            MAMMMXMMMM
            MXMXAXMASX
            """;
}